File: packages/galvanized-pukeko-agent-adk/src/main/java/io/github/galvanized_pukeko/agui/EventEncoder.java
package io.github.galvanized_pukeko.agui;

import com.agui.core.types.AgUiJsonKt;
import com.agui.core.types.BaseEvent;

/**
 * Vendored (Java reimpl) of the ag-ui {@code :kotlin-encoder} {@code EventEncoder} (BE-7). Swap for
 * {@code com.ag-ui.community:kotlin-encoder} when BE-3's upstream PR publishes.
 *
 * <p>Encodes AG-UI {@link BaseEvent}s to the wire form. Mirrors the TypeScript
 * ({@code @ag-ui/encoder}) and Python ({@code ag_ui.encoder}) {@code EventEncoder}: construct per
 * request with the client's {@code Accept} header, then {@link #encode(BaseEvent)} each event.
 * SSE only for now (no Kotlin proto codec yet).
 *
 * <p>Serialization goes through the same proven Java&harr;kotlinx idiom already used in this module
 * (see {@code AgUiController} decode side + {@code AgUiWireTest}): the {@code AgUiJson} top-level
 * Kotlin {@code val} is reached from Java as {@code AgUiJsonKt.getAgUiJson()}, and the polymorphic
 * {@link BaseEvent} is serialized with its generated {@code BaseEvent.Companion.serializer()}.
 */
public final class EventEncoder {

    public static final String SSE_CONTENT_TYPE = "text/event-stream";
    public static final String AGUI_MEDIA_TYPE = "application/vnd.ag-ui.event+proto";

    private final String accept;

    public EventEncoder(String accept) {
        this.accept = accept;
    }

    /** Always SSE for now (no Kotlin proto codec yet). */
    public String getContentType() {
        return SSE_CONTENT_TYPE;
    }

    /** Mirrors TS/Python {@code encode()}. */
    public String encode(BaseEvent event) {
        return encodeSSE(event);
    }

    /** Canonical SSE framing: {@code data: {json}\n\n}. */
    public String encodeSSE(BaseEvent event) {
        return "data: " + encodeToJson(event) + "\n\n";
    }

    /**
     * The JSON wire body for {@code event}, with <b>no</b> SSE framing. Use from transports that add
     * their own framing (Spring MVC's {@code SseEmitter}, which prefixes {@code data:} and appends the
     * blank line) — pairing those with {@link #encode(BaseEvent)} would double-frame the stream.
     */
    public String encodeToJson(BaseEvent event) {
        return AgUiJsonKt.getAgUiJson().encodeToString(BaseEvent.Companion.serializer(), event);
    }
}

File: examples/pukeko-koog-ag-ui/koog-agent/src/main/kotlin/com/pukeko/example/koog/EventEncoder.kt
// Vendored from ag-ui :kotlin-encoder (BE-7). Swap for com.ag-ui.community:kotlin-encoder when BE-3's upstream PR publishes.
package com.pukeko.example.koog

import com.agui.core.types.AgUiJson
import com.agui.core.types.BaseEvent
import kotlinx.serialization.encodeToString

/**
 * Encodes AG-UI [BaseEvent]s to the wire form. Mirrors the TypeScript
 * (`@ag-ui/encoder`) and Python (`ag_ui.encoder`) `EventEncoder`: construct per
 * request with the client's `Accept` header, then [encode] each event.
 *
 * SSE only for now (matching the Python encoder). Protobuf negotiation is a TODO
 * pending a Kotlin `@ag-ui/proto` codec — see [getContentType].
 */
class EventEncoder(private val accept: String? = null) {

    /** Mirrors TS/Python `getContentType()`. Always SSE for now (no Kotlin proto codec yet). */
    fun getContentType(): String = SSE_CONTENT_TYPE
    // TODO: negotiate AGUI_MEDIA_TYPE from `accept` once a Kotlin proto
    // codec exists, matching TS EventEncoder.getContentType()/encodeBinary().

    /** Mirrors TS/Python `encode()`. */
    fun encode(event: BaseEvent): String = encodeSSE(event)

    /** Canonical SSE framing: `data: {json}\n\n`. Mirrors TS `encodeSSE`. */
    fun encodeSSE(event: BaseEvent): String =
        "data: " + encodeToJson(event) + "\n\n"

    /**
     * The JSON wire body for [event], with **no** SSE framing.
     *
     * Use this from transports that add their own framing — e.g. Spring MVC's `SseEmitter` or
     * Jakarta's `SseEventSink`, which prefix `data:` and append the blank line themselves. Pairing
     * such a transport with [encode]/[encodeSSE] would double-frame the stream; hand it
     * [encodeToJson] instead so the SDK stays the single serialization entry point (no need for the
     * consumer to reach past the encoder to [AgUiJson]/`BaseEvent.serializer()` directly). For raw
     * byte streaming (FastAPI-style `StreamingResponse`, Ktor `respondBytesWriter`) use [encode].
     */
    fun encodeToJson(event: BaseEvent): String = AgUiJson.encodeToString(event)

    companion object {
        const val SSE_CONTENT_TYPE: String = "text/event-stream"
        const val AGUI_MEDIA_TYPE: String = "application/vnd.ag-ui.event+proto"
    }
}
